### DISCUSSION TITLE
Building runtime system which wraps main on top of HPX

### DISCUSSION THREAD

**Discussion by antoniupop** [STEllAR-GROUP/hpx#5673] [Q&A]
I'm looking at how to get the main wrapper used in HPX and the one I need to have in my own dataflow runtime system (let's call it DFR)  to work gracefully  along each other. Right now I am unable to get this to work at all (including using the compiler's main constructor/destructor). Have you seen other similar uses? Any idea of how (or if) this could work? Normally the HPX wrapper should be the outermost - the constructor/destructor allow both to coexist but the latter becomes outermost, so the DFR initialisation cannot issue calls for initialising HPX.

Thanks!

**Comment by hkaiser** [comment]
The easiest way of coordinating the initialization of HPX with your system startup might be to take things into your own hands and to do it explicitly. As long as you don't `#include <hpx/hpx_main.hpp>` we will not do anything to silently initialize HPX. That should give you all the options you need to reliably first initialize HPX and then your environment (or v.v. if needed).

**Reply by antoniupop** [reply]
I have tried that a while back, but had run into a number of issues in HPX, in particular getting the "component is disabled for this locality (component_invalid[-1]): HPX(bad_request)" error. I saw the discussion (https://irclog.cct.lsu.edu/ste~b~~b~ar/2020-10-15) about something that felt similar but didn't solve my problem so gave up on that option. 
I'm happy to give it a try again - right now what I use is `#include <hpx/hpx_start.hpp>` and initialising with `hpx::start(nullptr, argc, argv)`
but it still does use the __wrap_main I believe which locks me out of initialising in the proper order.

**Reply by msimberg** [reply]
That sounds odd. The only way in which `main` should be wrapped by HPX is if you're including `hpx/wrap_main.hpp` (or `hpx/hpx_main.hpp`) somewhere in your application and linking your application to `HPX::wrap_main` in CMake. Are you doing either of those? Also, which version of HPX are you working with?
